docs/madr/decisions/031-locality-aware-load-balancing.md

```diff
@@ -0,0 +1,420 @@
+          - to:
+              type: Any
+      overprovisioningFactor: 
+        percentage: 120                                     # (7)
+```
+
+
+(1) In `localZone` section, you configure the affinity of tags used for load balancing. If list is empty or localZone is not specified, load balancing works normally, without taking any priorities into account. 
+
+(2) In `crossZone` section, you configure zones priority.
+
+(3) In `failover` section, you configure list of zone priorities, first rules will have the highest load balancing priority and last in order will have the lowest priority.
+
+(4) In `from` section, you configure to which zones this rule applies. This rule will apply to all DPPs in `zone-1` and `zone-2`
+
+(5) In `to` section, you configure to which zones traffic should be redirected.
+
+(6) `type` field lets you control how fallback zones are picked. Allowed values: 
+- `Only` - will load balance traffic only to zones listed in `zones` field 
+- `None` - will not load balance traffic any further
+- `Any` - will load balance traffic to all zones
+- `AnyExcept` - will load balance traffic to all zones except ones specified in `zones` field
+
+(9) In [`overprovisioningFactor`](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/overprovisioning#arch-overview-load-balancing-overprovisioning-factor) section, you configure how early `Envoy` should consider other priorities.
+
+
+### API examples based on use cases
+
+#### Keeping traffic in same zone (disable cross zone traffic)
+
+```yaml
+targetRef:
+  kind: Mesh
+to:
+  targetRef:
+    kind: MeshService
+    name: backend
+  defaults:
+    localityAwareness:
+      localZone: 
+        affinityTags: []
+```
+
+![Use case 1](assets/031/use_case_1.png)
+
+#### Load balancing based on priorities within a zone. Example: node → AZ → region
+
+```yaml
+targetRef:
+  kind: Mesh
+to:
+  targetRef:
+    kind: MeshService
+    name: backend
+  defaults:
+    localityAwareness:
+      localZone:
+        affinityTags: ["k8s.io/node", "k8s.io/az"]
+```
+
+![Use case 2](assets/031/use_case_2_1.png)
+![Use case 2](assets/031/use_case_2_2.png)
+![Use case 2](assets/031/use_case_2_3.png)
+
+#### Prefer traffic in same zone with fallback to any other zone
+
+```yaml
+targetRef:
+  kind: Mesh
+to:
+  targetRef:
+    kind: MeshService
+    name: backend
+  defaults:
+    localityAwareness:
+      localZone:
+        affinityTags: []
+      crossZone:
+        failover:
+          - to: 
+              type: Any
+```
+
+![Use case 3](assets/031/use_case_3_1.png)
+![Use case 3](assets/031/use_case_3_2.png)
+
+#### Load balance equally to list of zones
+In this case, traffic from outside to the `backend` service in `zone-2` is completely excluded.
+
+```yaml
+targetRef:
+  kind: Mesh
+to:
+  targetRef:
+    kind: MeshService
+    name: backend
+  defaults:
+    localityAwareness:
+      localZone:
+        affinityTags: []
+      crossZone:
+        failover:
```

Comment:
The word "failover" implies the behavior in this section applies to situations when something went wrong (i.e. requests to local instance started failing and we're looking for an alternative). But the use case "Load balance equally to list of zones" means sending requests between local/zone-1/zone-3 even if everything works as expected

Reply:
I think the description is misleading. We always route to local first and if that fails we are going out of the local zone.

Maybe it should be:
```Load balance equally to list of zones when local not available````

Reply:
I think that the case, maybe in this case you should use Mesh*Route https://github.com/kumahq/kuma/pull/8022/files#diff-ce8ace61d35c2efd60262c423f5763d63fb047d8d7b000deefa90beb647bb36aR250

Reply:
Yeah it's also odd you would prefer more complex request routing if you can fullfill it locally.

Reply:
Also, what policy takes precedence? If I have MeshLoadBalancingStrategy that prioritizes Local, but then I created MeshHTTPRoute like:

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshHTTPRoute
metadata:
  name: route-all-zones
  namespace: kuma-system
  labels:
    kuma.io/mesh: default
spec:
  targetRef:
    kind: MeshService
    name: frontend
  to:
  - targetRef:
      kind: MeshService
      name: backend
    rules:
      default:
        backendRef:
        - kind: MeshServiceSubset
          name: backend
          tags:
            kuma.io/zone: zone-1
        - kind: MeshServiceSubset
          name: backend
          tags:
            kuma.io/zone: zone-3
```

What will be the behavior?

Reply:
@lobkovilya does this address your concerns?

Reply:
Order of plugins is `Mesh*Route` and after it `MeshLoadBalancingStrategy`
So when you create
```
apiVersion: kuma.io/v1alpha1
kind: MeshHTTPRoute
metadata:
  name: route-all-zones
  namespace: kuma-system
  labels:
    kuma.io/mesh: default
spec:
  targetRef:
    kind: MeshService
    name: frontend
  to:
  - targetRef:
      kind: MeshService
      name: backend
    rules:
      default:
        backendRef:
        - kind: MeshServiceSubset
          name: backend
          tags:
            kuma.io/zone: zone-1
        - kind: MeshServiceSubset
          name: backend
          tags:
            kuma.io/zone: zone-3
```

later `MeshLoadBalancingStrategy` takes endpoints and configures them based on the routing policy. So if your MLBS allows to route to these zones you will route there

Reply:
what MLBS I have to create to make `route-all-zones` MHR work as expected?

Reply:
I see, having MLBS like this
```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshLoadBalancingStrategy
metadata:
  name: mlbs-any
  namespace: kuma-system
  labels:
    kuma.io/mesh: default
spec:
  targetRef:
    kind: Mesh
  to:
    targetRef:
      kind: MeshService
      name: backend
    defaults:
      localityAwareness:
        crossZone:
          failover:
            - to:
                type: Any
```
allows me to create any possible split in MHR.